cli/command/formatter/sorter.go

```diff
@@ -0,0 +1,321 @@
+package formatter
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+)
+
+const (
+	ascendantKey  = "asc"
+	descendantKey = "desc"
+)
+
+type fieldIndex []int
+
+type orderBy struct {
+	name     string     // order by name
+	ordering string     // asc or desc
+	index    fieldIndex // index of field in struct
+}
+
+type genericStructSorter struct {
+	data []interface{}
+	bys  []orderBy
+}
+
+func (s genericStructSorter) Len() int {
+	return len(s.data)
+}
+
+func (s genericStructSorter) Swap(i, j int) {
+	s.data[i], s.data[j] = s.data[j], s.data[i]
+}
+
+// validate all the data are of same time
+func validateType(data []interface{}) error {
+	if len(data) == 0 {
+		return nil
+	}
+	baseType := reflect.TypeOf(data[0])
+	if baseType.Kind() != reflect.Struct {
+		return fmt.Errorf("data must be of type struct")
+	}
+	for i := 1; i < len(data); i++ {
+		iType := reflect.TypeOf(data[i])
+		if iType.PkgPath() != baseType.PkgPath() || iType.Name() != baseType.Name() || iType.Kind() != baseType.Kind() {
+			return fmt.Errorf("data are not of the same type: (%v, %v, %v) != (%v, %v, %v)",
+				iType.PkgPath(), iType.Name(), iType.Kind(), baseType.PkgPath(), baseType.Name(), baseType.Kind())
+		}
+	}
+	return nil
+}
+
+// newGenericStructSorter creates a sorter based on data and user specified configure
+// whitelist is a map with struct field name as key and new name as value
+func newGenericStructSorter(data []interface{}, bys []string, whitelist map[string]string) (genericStructSorter, error) {
+	gs := genericStructSorter{}
+	if len(data) == 0 {
+		return gs, nil
+	}
+	if len(bys) == 0 {
+		return gs, fmt.Errorf("bys cannot be empty")
```

Comment:
If we only create a sorter if bys is not empty, this check is not needed.

Also, in which situation would we get an empty list (and is it an actual _error_?) Could probably be just;

```go
if len(data) == 0 || len(bys) == 0 {
    return gs, nil
}
```